### ISSUE TITLE
Should object.playerId be object.ownerId in ClientEngine.isOwnedByPlayer(object)?

### ISSUE THREAD

**Issue by danielytics** [lance-gg/lance#57]
In the following code snippet, `object.playerId` is used to check if an object is owned by the player, but, wouldn't this only return true for the single player object since all other objects use `object.ownerId` for ownership?

https://github.com/lance-gg/lance/blob/78e91bc6a8ed62e44e1a0d95ea00480de72039fa/src/ClientEngine.js#L103-L105

**Comment by OpherV** [comment]
You have great timing - I was planning the refactoring of this bit yesterday. 
This is a good opportunity to open this issue up for discussion.

Lance in the latest stable version has no concept of 'ownership'; The first iteration of implementation started with `dynamicObject`s having a `playerId` which should correspond to the player that created them. 

However, while implementing Spaaace, I found this paradigm to be insufficient, since AIs don't have a playerId. We still want to check collisions between missiles and ships regardless if the ship is player or AI controlled. Thus the local implementation of `ownerId` in Spaaace was born - it only exists in Spaaace, and not in Lance. This is inelegant and I want to fix it for Lance V2.

There are several options for tackling this. The first is keeping `ownerId` in the netScheme of every `dynamicObject`. The major downside is that it means sending a lot of extra data across the wire, data that usually doesn't change at all. That's a waste of precious bandwidth.

Not all classes might need an owner. A 'power up' class in a game, that a character can pick up doesn't have an 'owner'. So another option is having an "ownership" component that only some classes will add (like the missile in the Spaaace example). Only these classes with this component will have ownerId in their netScheme - I think this API is a bit too cumbersome, and not intuitive

The third option is implementing some sort of 'constant' values that are sent across the wire on object creation. So for values that are set at creation time and don't change at all they will be sent only once. This touches on the topic of how to implement 'atomic' events. That is, events that always reach their destination even on lossy connections like UDP. So for example if the payload containing the 'ship created' event is lost, it is resent to the client and further events (position/health updates etc) are ignored. This is cool, but very non-trivial to implement

What do you think would be a convenient API for game designers to use?

**Comment by OpherV** [comment]
@danielytics after much thought I think the simplest approach might also be the best:
Instead of having both ownerId AND playerId, we could just have just playerId and assign it to AI players as well. 

It works out semantically, because those are still technically players, only that they are computer-controlled instead of human-controlled. That way there's no effect on networking, nor do we need any extra mapping between players and owners.

With this in mind the answer to the original question would be 'no' :)
Does this make sense?

**Comment by danielytics** [comment]
Does that mean that each NPC would effectively be a player with its own unique player id? Can there only be one gameobject per player id then (the above code returns the first object where the id matches)?

What about objects the are actually owned by a player? For example proximity mines that don't blow up for the player that placed them. I suppos Lance doesn't have to handle his use case as you can simply manually add an owner field for the objects that need one, but if it's a common use case it might be nice to have it built in (but it could be handled by a component once the ECS is implemented).

**Comment by OpherV** [comment]
No. There can be many gameObjects with the same playerId (and each with its own separate, unique id). In an RTS game, for example, all units of a certain player will have the same playerId whether it's an AI or a human player. This also covers the mine use case you specified. You can look at the spaaace as a degenerate case where every player has only one unit (the ship).

**Comment by danielytics** [comment]
Ok, so, `playerId` becomes what `ownerId` in the old spaaace game was. Is the unique Id a unique gameobject id? I assume then its up to the dev to determine if a concept of a player gameobject makes sense for their game or not (so for an action game with a player character, that characters unique id or a direct reference to the object, would be stored in ClientEngine; but for an RTS, there is no single player object at all).

**Comment by OpherV** [comment]
yes - every gameObject has a unique ID

Resolved in the dev branch (see example implementation in [simple-spaaace](https://github.com/OpherV/simple-spaaace)]
